### PR TITLE
Api token authentication

### PR DESCRIPTION
**Pull request by MancunianSam**

### CODE REVIEW

File: src/main/scala/http/Routes.scala [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import org.keycloak.adapters.rotation.AdapterTokenVerifier
+import spray.json.JsValue
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Try
+
+
+object Routes {
+
+  implicit val system: ActorSystem = ActorSystem("helloAkkaHttpServer")
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/main/scala/http/Routes.scala]: I've changed the name

File: src/main/scala/http/Routes.scala [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  implicit val executionContext: ExecutionContext = system.dispatcher
+
+
+  val ttl: Int = 60 * 10
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/main/scala/http/Routes.scala]: I've renamed it ttlSeconds.

File: src/main/scala/http/Routes.scala [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  val url: String = ConfigFactory.load().getString("auth.url")
+  val keycloakDeployment = TdrKeycloakDeployment(url, "tdr", ttl)
+
+
+  def verifyToken(token: String): Boolean =
+    Try(Option(AdapterTokenVerifier.verifyToken(token, keycloakDeployment)).isDefined).getOrElse(false)
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/main/scala/http/Routes.scala]: I've changed the method to warn if either the token is missing or expired.

File: src/main/scala/http/Routes.scala [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+  def tokenAuthenticator(credentials: Credentials): Future[Option[String]] = {
+    credentials match {
+      case Credentials.Provided(token) => Future {
+        Some(token).filter(verifyToken)
+      }
+      case _ => Future.successful(None)
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/main/scala/http/Routes.scala]: It returns None if the filter is false and then Future.successful(None) translates to 401

File: src/main/scala/http/TdrKeycloakDeployment.scala [hunk 5]
```diff
@@ -0,0 +1,27 @@
+package http
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/main/scala/http/TdrKeycloakDeployment.scala]: I've changed it to uk.gov.nationalarchives.tdr.api

File: src/main/scala/http/TdrKeycloakDeployment.scala [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+class TdrKeycloakDeployment extends KeycloakDeployment {
+}
+
+object TdrKeycloakDeployment {
+  def apply(authServer: String, realm: String, ttl: Int): TdrKeycloakDeployment = {
+    val keycloakDeployment = new TdrKeycloakDeployment()
+    keycloakDeployment.setClient(new HttpClientBuilder().build())
+
+    val ttl = 60 * 10
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/main/scala/http/TdrKeycloakDeployment.scala]: Yes I did, thanks :+1:

File: src/test/resources/scripts/init.sql [hunk 7]
```diff
@@ -0,0 +1,10 @@
+CREATE SCHEMA IF NOT EXISTS consignmentapi;
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/test/resources/scripts/init.sql]: Not really, It should be ok for now I think and we can revisit it later.

File: src/test/scala/routes/SeriesRouteSpec.scala [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+import org.scalatest.matchers.should.Matchers
+import utils.TestUtils.validToken
+
+class SeriesRouteSpec extends AnyFlatSpec with Matchers with ScalatestRouteTest with BeforeAndAfterEach  {
+
+  override def beforeEach(): Unit = {
+    DbConnection.db.source.createConnection().prepareStatement("delete from consignmentapi.Series").executeUpdate()
+  }
+
+  "The api " should "return an empty series list" in {
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/test/scala/routes/SeriesRouteSpec.scala]: I must have copied and pasted each of these. I've taken the space out.

File: src/test/scala/routes/SeriesRouteSpec.scala [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    val ps: PreparedStatement = DbConnection.db.source.createConnection().prepareStatement("insert into consignmentapi.Series (SeriesId) VALUES (1)")
+    ps.executeUpdate()
+    val query: String = """{"query":"{getSeries{seriesid}}"}"""
+    Post("/graphql").withEntity(ContentTypes.`application/json`, query) ~> addCredentials(validToken) ~> route ~> check {
+      responseAs[String] shouldEqual "{\"data\":{\"getSeries\":[{\"seriesid\":1}]}}"
+    }
+  }
+
+  "The api " should "return all requested fields" in {
+    val sql = "insert into consignmentapi.Series (SeriesId, BodyId, Name, Code, Description) VALUES (1,1,'Name','Code','Description')"
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/test/scala/routes/SeriesRouteSpec.scala]: This is all just an idea on how the routes testing works. We can always change any bit if it's not working.

File: src/test/scala/routes/RouteAuthenticationSpec.scala [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+import akka.http.scaladsl.server.AuthenticationFailedRejection.{CredentialsMissing, CredentialsRejected}
+import akka.http.scaladsl.testkit.ScalatestRouteTest
+import http.Routes.route
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+import utils.TestUtils.{validToken,invalidToken}
+
+class RouteAuthenticationSpec extends AnyFlatSpec with Matchers with ScalatestRouteTest {
+
+  "The api " should "return ok" in {
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/test/scala/routes/RouteAuthenticationSpec.scala]: I've changed it to use healthcheck. There's a small PR here https://github.com/nationalarchives/tdr-terraform-environments/pull/31 to change it on the load balancer as well.

File: src/main/scala/http/Routes.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (suzannehamilton) [reply src/main/scala/http/Routes.scala]: Cheers. Would you mind renaming the parameter in TdrKeycloakDeployment for the same reason?

**Pull request merged**